sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/StorageBlockingSink.java

```diff
@@ -0,0 +1,106 @@
+ * implementing a BlockingSink.
+ */
+public class StorageBlockingSink {
+    private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+    /*
+    Note: Though this is used only by BlockBlobOutputStream, the decision was made to abstract this type out to make
+    testing easier.
+    */
+    private final Sinks.Many<ByteBuffer> writeSink;
+    private final LinkedBlockingQueue<ByteBuffer> writeLimitQueue;
+
+    /* Overrided implementation of LinkedBlockingQueue to effectively implement a true BlockingSink. */
+    private static class ProducerBlockingQueue<T> extends LinkedBlockingQueue<T> {
+        private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+        ProducerBlockingQueue(int queueSize) {
+            super(queueSize);
+        }
+
+        /* The Reactor implementation of Sinks(UnicastProcessor).tryEmitNext (see usage below) makes a call to
+        queue.offer(). The desired functionality of tryEmitNext is for it to block on backpressure until there is
+        space in the queue. Since this code is called from user thread on BlobOutputStream.write and blocking on that
+        API until data can be written is ok, this is the simplest way to get desired behavior. */
+        @Override
+        public boolean offer(T o) {
+            try {
+                super.put(o);
+                return true;
+            } catch (InterruptedException e) {
+                throw logger.logExceptionAsError(new RuntimeException(e));
+            }
+        }
+    }
+
+    /**
+     * Create a new StorageBlockingSink.
+     */
+    public StorageBlockingSink() {
+        /*
+        We use a LinkedBlockingQueue to effectively block the thread that calls tryEmitNext if there is
+        backpressure from downstream. Its capacity is 1 to keep the buffer as small as possible, as downstream
+        implementations do their own buffering.
+        */
+        this.writeLimitQueue = new ProducerBlockingQueue<>(1);
+        this.writeSink = Sinks.many().unicast().onBackpressureBuffer(writeLimitQueue);
+    }
+
+    /**
+     * Try to emit an element.
+     *
+     * @param buffer {@link ByteBuffer} to emit.
+     * @throws IllegalStateException if an unrecoverable error was thrown.
+     */
+    public void tryEmitNext(ByteBuffer buffer) throws IllegalStateException {
+        /*
+        tryEmitNext returns a Sinks.EmitResult that indicates different success/error cases we can
+        potentially handle (in case we want to retry transmission).
+        */
+        final Sinks.EmitResult writeResult = this.writeSink.tryEmitNext(buffer);
+        switch (writeResult) {
+            case OK: // Success
+                return;
+            case FAIL_OVERFLOW: /* When queue overflows. This indicates there is backpressure.
+            NOTE: If the FAIL_OVERFLOW case ever gets hit, it indicates there is a mismatch between the way Reactor
```

Comment:
Should I add an explicit error here for FAIL_OVERFLOW?
Or an explicit assert

Reply:
Related -> Should we just go back to using tryEmitNext().orThrow()?

Reply:
` tryEmitNext().orThrow()` sounds good if we keep how interruptedexception is handled